source/extensions/common/tap/admin.cc

```diff
@@ -110,40 +109,213 @@ void AdminHandler::unregisterConfig(ExtensionConfig& config) {
+    std::vector<envoy::data::tap::v3::TraceWrapper> buffer = trace_buffer->flush();
+
+    ENVOY_LOG(debug, "admin writing buffered trace list to response");
+
+    // Serialize writes to the stream
+    for (size_t i = 0; i < buffer.size(); i++) {
+      // Close stream on final message
+      attached_request->streamMsg(buffer[i], (i + 1) == buffer.size());
     }
+  });
+}
+
+AdminHandler::AttachedRequest::AttachedRequest(AdminHandler* admin_handler,
+                                               const envoy::admin::v3::TapRequest& tap_request,
+                                               Server::AdminStream* admin_stream)
+    : config_id_(tap_request.config_id()), config_(tap_request.tap_config()),
+      admin_stream_(admin_stream), main_thread_dispatcher_(admin_handler->main_thread_dispatcher_) {
+}
 
-    ENVOY_LOG(debug, "admin writing buffered trace to response");
-    Buffer::OwnedImpl output_buffer{output_string};
-    parent.attached_request_.value().admin_stream_->getDecoderFilterCallbacks().encodeData(
-        output_buffer, false);
+AdminHandler::AttachedRequestBuffered::AttachedRequestBuffered(
+    AdminHandler* admin_handler, const envoy::admin::v3::TapRequest& tap_request,
+    Server::AdminStream* admin_stream)
+    : AttachedRequest(admin_handler, tap_request, admin_stream) {
+  const envoy::config::tap::v3::OutputSink& sink =
+      tap_request.tap_config().output_config().sinks(0);
+
+  const uint64_t max_buffered_traces = sink.buffered_admin().max_traces();
+  const uint64_t timeout_ms = PROTOBUF_GET_MS_OR_DEFAULT(sink.buffered_admin(), timeout, 0);
+  trace_buffer_ = std::make_unique<TraceBuffer>(max_buffered_traces);
+  // Start the countdown if provided an actual timeout
+  if (timeout_ms > 0) {
+    timer_ = dispatcher().createTimer(
+        [this, admin_handler] { this->onTimeout(admin_handler->attached_request_); });
+    timer_->enableTimer(std::chrono::milliseconds(timeout_ms));
+  }
+}
+
+void AdminHandler::AttachedRequestBuffered::onTimeout(
+    const std::weak_ptr<AttachedRequest>& weak_attached_request) {
+  // Flush the buffer regardless of size
+  dispatcher().post([weak_attached_request] {
+    // Take temporary ownership - extend lifetime
+    std::shared_ptr<AttachedRequest> attached_request = weak_attached_request.lock();
+    if (!attached_request) {
+      // NOTE: Cannot do much here in response to the failed post as an HTTP response code has
+      // already been sent on completion of AdminHandler::handler. Additionally we probably don't
+      // want to take any action in this event as this case may be hit in "normal" usage, depending
+      // on when the destruction of attached_request_ occurs.
+      ENVOY_LOG(debug, "Timer Expiry after admin tap request completion");
+      return; // No attached request, flushed already.
+    }
+    TraceBuffer* trace_buffer = attached_request->traceBuffer();
+
+    // if the trace buffer has already been flushed short circuit
```

Comment:
How does this happen if everything is serialized? More comments?

Reply:
This case can be hit in the following scenario:
1. The final trace needed to fill the buffer arrives and `submitTrace` posts to the dispatcher.
2. The timeout expires and the on timeout callback is posted to the dispatcher.
3. At this point we have the dispatcher queue with a `submitTrace` post followed by an on timeout post.
4. The callback posted in step 1 runs which flushes the buffer and, posts the on destroy callback for the admin stream. At this point the buffer is flushed but the attached request has not been destroyed.
5. The timeout callback runs, triggering this case.

See `admin_test.cc:BufferedTapDoubleFlush` for the test which covers this case.

Reply:
Yup OK, can you add more comments? Same in the other case?